accepted/2022/FunctionPointers.md

```diff
@@ -0,0 +1,431 @@
+
+# Potential future features
+Various features not planned for 8.0; somewhat discussed earlier.
+
+## Support `Type.FullName` and `Type.AssemblyQualifiedName`
+Including `Type.GetType()` and similar APIs to parse that string to construct a function pointer type dynamically.
+
+## `System.Reflection.FunctionPointer` for `System.Object`-based invoke
+See the existing [System.Reflection.Pointer](https://docs.microsoft.com/dotnet/api/system.reflection.pointer) for reference. This would allow passing strongly-typed, validated function pointers when using `System.Object`-based invoke. Naming of Box\Unbox vs. constructors etc. is TBD.
+
+ This is a value type, not a reference type like `Pointer`, since newer proposed reflection APIs will not force boxing to occur.
+
+ Like `Pointer` there is no property to obtain the type since it is used in a temporary manner to manual marshal.
+
+```diff
+namespace System.Reflection
+{
++   public struct FunctionPointer
++   {
++       public static FunctionPointer Box(void* ptr, Type type);
++       public static void* Unbox (FunctionPointer ptr);
++   }
+}
+```
+
+## System.Reflection.MethodSignature
+Currently this is called "MethodSignature" with the intention that this may also support `MethodBase` and\or `Delegate` in the future.  If that was not the case, this would be called "FunctionPointerSignature".
+
+Shown is support for a late-bound invoke analogous to [Delegate.DynamicInvoke](https://docs.microsoft.com/dotnet/api/system.delegate.dynamicinvoke) and [MethodBase.Invoke](https://docs.microsoft.com/dotnet/api/system.reflection.methodbase.invoke).
+
+We may also want to add an overload to `ILGenerator.EmitCalli()` to take a "MethodSignature".
+
+`Type` exposes various methods to construct pointers, arrays and generics through `Type.MakePointerType()` etc. However, there is no proposal for V8 to add similar methods to function pointers. The "MethodSignature" class would assist with that, and could either be passed as-is to represent function pointer metadata, or we could expose a `Type CreateFunctionPointerType()` method.
+
+```diff
+namespace System.Reflection
+{
++   public sealed class MethodSignature
++   {
++       public MethodSignature(Type functionPointerType);
++       public MethodSignature(Type returnType, MethodParameter[]? parameters = null, Type[]? callingConventions = null);
++       public MethodParameter ReturnType { get; }
++       public MethodParameter[] Parameters { get; }
++       public Type[] CallingConventions { get; }
++       public bool IsUnmanaged { get; }
+
+        // We could expose this mechanism if useful.
++       public Type CreateFunctionPointerType();
+
+        // MethodInvoker class defined elsewhere
+        // Throws InvalidOperation when used with MetadataLoadContext types
++       public MethodInvoker Invoker { get; }
++   }
+
++   public sealed class MethodParameter
```

Comment:
This looks like a subset of ParameterInfo.

Reply:
I added a comment that we might be able to use ParameterInfo. This was also somewhat discussed when we had "FunctionPointerParameterInfo" and we decided to just create a new type instead of re-using ParameterInfo.